contracts/services/forwarder/src/tests.rs

```diff
@@ -98,192 +48,105 @@ impl ForwarderTestSuite {
             crate::contract::query,
         );
 
-        let forwarder_code_id = app.store_code(Box::new(forwarder_code));
+        let forwarder_code_id = inner.app_mut().store_code(Box::new(forwarder_code));
 
         Self {
-            app,
-            owner,
-            processor,
-            account_code_id,
-            account_salt: salt,
+            inner,
             forwarder_code_id,
             input_addr,
             output_addr,
+            input_balances,
         }
     }
 
-    fn api(&self) -> &MockApi {
-        self.app.api()
-    }
-
-    fn instantiate(&mut self, svc_cfg: &ServiceConfig) -> AnyResult<Addr> {
-        let msg = InstantiateMsg {
-            owner: self.owner.to_string(),
-            processor: self.processor.to_string(),
-            config: svc_cfg,
+    pub fn forwarder_init(&mut self, cfg: &ServiceConfig) -> Addr {
+        let init_msg = InstantiateMsg {
+            owner: self.owner().to_string(),
+            processor: self.processor().to_string(),
+            config: cfg.clone(),
         };
+        let addr = self.contract_init(self.forwarder_code_id, "forwarder", &init_msg, &[]);
+
+        let input_addr = self.input_addr().clone();
+        if self.inner.app_mut().contract_data(&input_addr).is_err() {
+            let account_addr = self.account_init("input_account", vec![addr.to_string()]);
+            assert_eq!(account_addr, input_addr);
+
+            if let Some(balances) = self.input_balances.as_ref().cloned() {
+                let amounts = balances
+                    .iter()
+                    .map(|(amount, denom)| coin(*amount, denom.to_string()))
+                    .collect::<Vec<Coin>>();
+                self.init_balance(&input_addr, amounts);
+            }
+        }
```

Comment:
Can you explain a little about this?

Reply:
Create the input account if it hasn't been created yet, and initialise its balance. The initial check with `contract_data` ensures we only do this once even if the function is mistakenly called multiple times.